File: scripts/provision_admin.js
/**
 * ShareByte Supabase Admin Provisioning Script (NodeJS)
 * ======================================================
 * Use this script with your SUPABASE_SERVICE_ROLE_KEY to provision a real Admin email.
 *
 * Environment variables:
 *   SUPABASE_URL=https://<your-project-id>.supabase.co
 *   SUPABASE_SERVICE_ROLE_KEY=eyJhbGciOiJKV1Qi...
 *
 * Usage:
 *   node scripts/provision_admin.js indumedagam@gmail.com "Indu Admin"
 */

const { createClient } = require('@supabase/supabase-js');

const supabaseUrl = process.env.SUPABASE_URL || process.env.VITE_SUPABASE_URL;
const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !serviceRoleKey) {
  console.log('⚠️ SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables required.');
  console.log('Usage: SUPABASE_URL=... SUPABASE_SERVICE_ROLE_KEY=... node scripts/provision_admin.js <admin_email>');
  process.exit(1);
}

const adminEmail = process.argv[2];
const adminName = process.argv[3] || 'System Administrator';

if (!adminEmail || !adminEmail.includes('@')) {
  console.error('❌ Error: Please provide a valid admin email address.');
  process.exit(1);
}

const supabaseAdmin = createClient(supabaseUrl, serviceRoleKey, {
  auth: { autoRefreshToken: false, persistSession: false }
});

async function provisionAdminAccount() {
  console.log(`🔍 Provisioning ADMIN account for ${adminEmail}...`);

  // 1. Create or fetch Auth User
  const { data: authUser, error: authErr } = await supabaseAdmin.auth.admin.createUser({
    email: adminEmail,
    email_confirm: true,
    user_metadata: { full_name: adminName }
  });

  let userId;
  if (authErr) {
    if (authErr.message.includes('already exists')) {
      console.log('ℹ️ Auth user already exists. Fetching user profile...');
      const { data: usersData } = await supabaseAdmin.auth.admin.listUsers();
      const userObj = usersData.users.find(u => u.email === adminEmail);
      if (!userObj) {
        console.error('❌ Error locating existing user.');
        process.exit(1);
      }
      userId = userObj.id;
    } else {
      console.error('❌ Auth error:', authErr.message);
      process.exit(1);
    }
  } else {
    userId = authUser.user.id;
  }

  // 2. Upsert profile in PostgreSQL `profiles` table with role='ADMIN' & account_status='ACTIVE'
  const { error: profileErr } = await supabaseAdmin
    .from('profiles')
    .upsert({
      id: userId,
      email: adminEmail,
      full_name: adminName,
      role: 'ADMIN',
      account_status: 'ACTIVE',
      email_verified: true,
      updated_at: new Date().toISOString()
    });

  if (profileErr) {
    console.error('❌ Error creating profiles record:', profileErr.message);
    process.exit(1);
  }

  console.log(`✅ SUCCESS: Admin account '${adminEmail}' successfully provisioned as ADMIN (ACTIVE).`);
}

provisionAdminAccount();

File: backend/scripts/provision_admin.py
"""
ShareByte Admin Provisioning Script
====================================
Use this script to securely provision a real Admin account in your database/Supabase system.

Usage:
  py -3 backend/scripts/provision_admin.py <admin_email> [full_name]

Example:
  py -3 backend/scripts/provision_admin.py indumedagam@gmail.com "Indu Admin"
"""

import sys
import os

# Add parent backend directory to sys.path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from app import create_app, db
from models import User

def provision_admin(email, full_name="System Administrator"):
    email = email.strip().lower()
    if not email or '@' not in email:
        print("❌ Error: Please provide a valid admin email address.")
        sys.exit(1)

    app = create_app()
    with app.app_context():
        existing = User.query.filter_by(email=email).first()
        if existing:
            print(f"ℹ️ User with email '{email}' already exists. Updating role to 'admin' and status to 'active'...")
            existing.role = 'admin'
            existing.status = 'active'
            existing.is_email_verified = True
            db.session.commit()
            print(f"✅ SUCCESS: Account '{email}' is now provisioned as ADMIN (active).")
        else:
            print(f"➕ Creating new provisioned ADMIN account for '{email}'...")
            new_admin = User(
                username=email,
                email=email,
                role='admin',
                status='active',
                is_email_verified=True,
                address='Central Admin Center'
            )
            new_admin.set_password('AdminSecureDefault123!')
            db.session.add(new_admin)
            db.session.commit()
            print(f"✅ SUCCESS: Provisioned ADMIN account created for '{email}'.")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("❌ Error: Missing admin email argument.")
        print("Usage: py -3 backend/scripts/provision_admin.py <admin_email> [full_name]")
        sys.exit(1)
        
    admin_email = sys.argv[1]
    name = sys.argv[2] if len(sys.argv) > 2 else "System Administrator"
    provision_admin(admin_email, name)
